File: Golang_src/easy/mergeSortedArray_88.go
package easy

// Use two pointers:
// Time Complexity: O(m+n)
// Space Complexity:O(1)
// Runtime: 0 ms, faster than 100.00%
// Memory Usage: 2.3 MB, less than 73.11%
func merge(nums1 []int, m int, nums2 []int, n int) {
	index1, index2 := m-1, n-1
	indexCur := m + n - 1

	for index1 >= 0 && index2 >= 0 {
		if nums1[index1] > nums2[index2] {
			nums1[indexCur] = nums1[index1]
			index1--
		} else {
			nums1[indexCur] = nums2[index2]
			index2--
		}
		indexCur--
	}

	for index2 >= 0 {
		nums1[indexCur] = nums2[index2]
		index2--
		indexCur--
	}
}

File: Cpp_src/easy/merge_sorted_array_88.cc
#include <iostream>
#include <vector>
using namespace std;

// Use conquer strategy without addition array:
// Time Complexity: O(m+n)
// Space Complexity:O(1)
// Runtime: 4 ms, faster than 85.29%
// Memory Usage: 8.8 MB, less than 65.22%
class Solution
{
public:
    void merge(vector<int> &nums1, int m, vector<int> &nums2, int n)
    {
        int index_1 = m - 1;
        int index_2 = n - 1;
        int index_cur = m + n - 1;

        while (index_1 >= 0 && index_2 >= 0)
        {
            if (nums1[index_1] > nums2[index_2])
            {
                nums1[index_cur--] = nums1[index_1--];
            }
            else
            {
                nums1[index_cur--] = nums2[index_2--];
            }
        }
        while (index_2 >= 0)
        {
            nums1[index_cur--] = nums2[index_2--];
        }
    }
};